satellite/satellitedb/nodeselection.go

```diff
@@ -0,0 +1,323 @@
+			}
+		}
+
+		if needNewNodes > 0 && receivedNewNodes == 0 {
+			needReputableNodes += needNewNodes
+			needNewNodes = 0
+		}
+
+		if needReputableNodes <= 0 && needNewNodes <= 0 {
+			break
+		}
+	}
+	return nodes, nil
+}
+
+func (cache *overlaycache) selectStorageNodesOnce(ctx context.Context, reputableNodeCount, newNodeCount int, criteria *overlay.NodeCriteria, excludedIDs []storj.NodeID, excludedNetworks []string) (reputableNodes, newNodes []*overlay.SelectedNode, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	newNodeSelection := `SELECT last_net, id, address, last_ip_port, true FROM nodes`
+	reputableNodeSelection := `SELECT last_net, id, address, last_ip_port, false FROM nodes`
+
+	if criteria.DistinctIP {
+		newNodeSelection = `SELECT DISTINCT ON (last_net) last_net, id, address, last_ip_port, true FROM nodes`
+		reputableNodeSelection = `SELECT DISTINCT ON (last_net) last_net, id, address, last_ip_port, false FROM nodes`
+	}
+
+	newNodesCondition, err := buildConditions(ctx, criteria, excludedIDs, excludedNetworks, true)
+	if err != nil {
+		return nil, nil, err
+	}
+	reputableNodesCondition, err := buildConditions(ctx, criteria, excludedIDs, excludedNetworks, false)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var newNodeQuery, reputableNodeQuery query
+	newNodeQuery = assemble(partialQuery{selection: newNodeSelection, condition: newNodesCondition, limit: newNodeCount})
+	if criteria.DistinctIP {
+		newNodeQuery = assemble(partialQuery{selection: newNodeSelection, condition: newNodesCondition, orderBy: "last_net", distinct: true, limit: newNodeCount})
+		newNodeQuery = selectRandomFrom(newNodeQuery, "filteredcandidates", newNodeCount)
+	}
+
+	reputableNodeQuery = assemble(partialQuery{selection: reputableNodeSelection, condition: reputableNodesCondition, limit: reputableNodeCount})
+	if criteria.DistinctIP {
+		reputableNodeQuery = assemble(partialQuery{selection: reputableNodeSelection, condition: reputableNodesCondition, orderBy: "last_net", distinct: true, limit: reputableNodeCount})
+		reputableNodeQuery = selectRandomFrom(reputableNodeQuery, "filteredcandidates", reputableNodeCount)
+	}
+
+	query := union(newNodeQuery, reputableNodeQuery)
+
+	rows, err := cache.db.Query(ctx, cache.db.Rebind(query.query), query.args...)
+	if err != nil {
+		return nil, nil, Error.Wrap(err)
+	}
+	defer func() { err = errs.Combine(err, rows.Close()) }()
+
+	for rows.Next() {
+		var node overlay.SelectedNode
+		node.Address = &pb.NodeAddress{Transport: pb.NodeTransport_TCP_TLS_GRPC}
+		var lastIPPort sql.NullString
+		var isNew bool
+		err = rows.Scan(&node.LastNet, &node.ID, &node.Address.Address, &node.LastIPPort, &isNew)
+		if err != nil {
+			return nil, nil, err
+		}
+		if lastIPPort.Valid {
+			node.LastIPPort = lastIPPort.String
+		}
+		if isNew {
+			newNodes = append(newNodes, &node)
+		} else {
+			reputableNodes = append(reputableNodes, &node)
+		}
+		if len(newNodes)+len(reputableNodes) == reputableNodeCount+newNodeCount {
```

Comment:
I would prefer `if len(newNodes) >= newNodeCount && len(reputableNodes) >= reputableNodeCount`

Reply:
okay cool I can change that